fdbserver/BlobManager.actor.cpp

```diff
@@ -1314,6 +1320,585 @@ ACTOR Future<Void> blobWorkerRecruiter(
+
+	state KeyRangeMap<UID>::iterator activeRange;
+	for (activeRange = activeRanges.begin(); activeRange != activeRanges.end(); ++activeRange) {
+		if (BM_DEBUG) {
+			printf("Checking if active range [%s-%s), owned by BW %s, should be pruned\n",
+			       activeRange.begin().printable().c_str(),
+			       activeRange.end().printable().c_str(),
+			       activeRange.value().toString().c_str());
+		}
+
+		// assumption: prune boundaries must respect granule boundaries
+		if (activeRange.begin() < startKey || activeRange.end() > endKey) {
+			continue;
+		}
+
+		// TODO: if this is a force prune, then revoke the assignment from the corresponding BW first
+		// so that it doesn't try to interact with the granule (i.e. force it to give up gLock).
+		// we'll need some way to ack that the revoke was successful
+
+		loop {
+			try {
+				if (BM_DEBUG) {
+					printf("Fetching latest history entry for range [%s-%s)\n",
+					       activeRange.begin().printable().c_str(),
+					       activeRange.end().printable().c_str());
+				}
+				Optional<GranuleHistory> history = wait(getLatestGranuleHistory(&tr, activeRange.range()));
+				// TODO: can we tell from the krm that this range is not valid, so that we don't need to do a get
+				if (history.present()) {
+					if (BM_DEBUG) {
+						printf("Adding range to history queue\n");
+					}
+					historyEntryQueue.push({ activeRange.range(), history.get().version, MAX_VERSION });
+				}
+				break;
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+	}
+
+	if (BM_DEBUG) {
+		printf("Beginning BFS traversal of history\n");
+	}
+	while (!historyEntryQueue.empty()) {
+		// process the node at the front of the queue and remove it
+		KeyRange currRange;
+		state Version startVersion;
+		state Version endVersion;
+		std::tie(currRange, startVersion, endVersion) = historyEntryQueue.front();
+		historyEntryQueue.pop();
+
+		if (BM_DEBUG) {
+			printf("Processing history node [%s-%s) with versions [%lld, %lld)\n",
+			       currRange.begin.printable().c_str(),
+			       currRange.end.printable().c_str(),
+			       startVersion,
+			       endVersion);
+		}
+
+		// get the persisted history entry for this granule
+		state Standalone<BlobGranuleHistoryValue> currHistoryNode;
+		state KeyRef historyKey = blobGranuleHistoryKeyFor(currRange, startVersion);
+		loop {
+			try {
+				Optional<Value> persistedHistory = wait(tr.get(historyKey));
+				ASSERT(persistedHistory.present());
+				currHistoryNode = decodeBlobGranuleHistoryValue(persistedHistory.get());
+				break;
+			} catch (Error& e) {
+				wait(tr.onError(e));
+			}
+		}
+
+		if (BM_DEBUG) {
+			printf("Found history entry for this node. It's granuleID is %s\n",
+			       currHistoryNode.granuleID.toString().c_str());
+		}
+
+		// if we already saw this node, skip it; otherwise, mark it as visited
+		// TODO: doing the visited check here as opposed to when adding this node as a parent
```

Comment:
Two possible approaches:
- make `std::pair<KeyRange, Version>` hashable (since `range, startVersion` can uniquely identify a granule)
- or, persist the granuleID for each parent as part of `parentGranules`

Reply:
Agreed

Reply:
Done